src/handle_timeout.cxx

```diff
@@ -102,7 +102,8 @@ void raft_server::handle_hb_timeout(int32 srv_id) {
 
 void raft_server::restart_election_timer() {
     // don't start the election timer while this server is still catching up the logs
-    if (catching_up_) {
+    // or this server is the leader
+    if (catching_up_ || role_ == srv_role::leader) {
```

Comment:
Would you please add `recur_lock(lock_);` at the beginning of this function? So far this function was internal use only, which means all callers already acquire `lock_` before reach here. But now we expose it to users so that this function should protect itself.